### PR TITLE
Implement scheduling in timer IRQ.

### PR DESCRIPTION
**Pull request by too-r**

### CODE REVIEW

**Comment by robert-w-gries** [comment]
Thanks @too-r!

Could you please add the following to prevent deadlocks during process creation:

```diff
diff --git a/src/arch/x86/interrupts/mod.rs b/src/arch/x86/interrupts/mod.rs
index b0a295d..fa93dd1 100644
--- a/src/arch/x86/interrupts/mod.rs
+++ b/src/arch/x86/interrupts/mod.rs
@@ -5,17 +5,18 @@ pub mod syscall;
 pub const DOUBLE_FAULT_IST_INDEX: usize = 0;
 
 /// Disable interrupts, execute code uninterrupted, re-enable interrupts
-pub fn disable_interrupts_then<T>(uninterrupted_fn: T) -> ()
+pub fn disable_interrupts_then<F, T>(uninterrupted_fn: F) -> T
 where
-    T: FnOnce() -> (),
+    F: FnOnce() -> T,
 {
     unsafe {
         ::x86::shared::irq::disable();
     }
-    uninterrupted_fn();
+    let result: T = uninterrupted_fn();
     unsafe {
         ::x86::shared::irq::enable();
     }
+    result
 }
 
 #[cfg(test)]
diff --git a/src/syscall/process.rs b/src/syscall/process.rs
index ddf5e89..75c2a0f 100644
--- a/src/syscall/process.rs
+++ b/src/syscall/process.rs
@@ -3,9 +3,13 @@ use scheduling::{DoesScheduling, ProcessId, SCHEDULER};
 
 /// Wrapper around scheduler.create() and ready() that can be called in processes
 pub fn create(new_proc: extern "C" fn(), name: String) -> ProcessId {
-    let pid = SCHEDULER
-        .create(new_proc, name)
-        .expect("Could not create new process!");
-    SCHEDULER.ready(pid.clone());
-    pid
+    use arch::interrupts;
+
+    interrupts::disable_interrupts_then(|| -> ProcessId {
+        let pid = SCHEDULER
+            .create(new_proc, name)
+            .expect("Could not create new process!");
+        SCHEDULER.ready(pid.clone());
+        pid
+    })
 }
```

File: src/arch/x86/interrupts/irq.rs [hunk 1]
```diff
@@ -15,7 +18,19 @@ fn trigger(irq: u8) {
...
 pub extern "x86-interrupt" fn timer(_stack_frame: &mut ExceptionStack) {
+    use arch::x86::interrupts;
+
     pic::MASTER.lock().ack();
+
+    if PIT_TICKS.fetch_add(1, Ordering::SeqCst) >= 10 {
+        PIT_TICKS.store(0, Ordering::SeqCst);
+
+        unsafe {
+            interrupts::disable_interrupts_then(|| {
```
Comment (robert-w-gries) [inline src/arch/x86/interrupts/irq.rs]: It seems a little more safe to include our manipulation of `PIT_TICKS` in the `disable_interrupts_then` closure.

I'm thinking something like this:

```diff
-    if PIT_TICKS.fetch_add(1, Ordering::SeqCst) >= 10 {
-        PIT_TICKS.store(0, Ordering::SeqCst);
+    interrupts::disable_interrupts_then(|| {
+        if PIT_TICKS.fetch_add(1, Ordering::SeqCst) >= 10 {
+            PIT_TICKS.store(0, Ordering::SeqCst);
 
-        unsafe {
-            interrupts::disable_interrupts_then(|| {
+            unsafe {
                 SCHEDULER.resched();
-            });
+            }
         }
-    }
+    });
```

File: src/arch/x86/interrupts/irq.rs [hunk 2]
```diff
@@ -15,7 +18,19 @@ fn trigger(irq: u8) {
 }
 
 pub extern "x86-interrupt" fn timer(_stack_frame: &mut ExceptionStack) {
+    use arch::x86::interrupts;
+
     pic::MASTER.lock().ack();
+
+    if PIT_TICKS.fetch_add(1, Ordering::SeqCst) >= 10 {
```
Comment (robert-w-gries) [inline src/arch/x86/interrupts/irq.rs]: Can you add a comment that notes how many milliseconds 10 `PIT_TICKS` would represent?

**Comment by robert-w-gries** [comment]
So there are still some deadlocks with the code as-is.  However, I cloned your fork, merged in the changes from #41, and added various `disable_interrupts_then` wrappers in critical sections of the kernel.

I am very pleased with the end result.  Your timer irq handler tests the scheduling module more thoroughly than any test code I had written for the initial PR.  I am now much more confident in my scheduling code.  The PIT irq handler is more important than I had realized.

I aim to merge your PR mostly as-is (pending a few changes) and patch in my additional work into the #41 PR.  I think this is the best way to get your code merged in without bothering you too much 😄

**Comment by too-r** [comment]
I'm fine to make these changes and then we can merge this as one rather than making extra changes elsewhere

**Comment by too-r** [comment]
I don't think it is really necessary to manipulate PIT_TICKS within the closure, since it is not inherently related to the irq - it is just a counter variable.

**Comment by too-r** [comment]
I've fixed some basic syntax errors. This now builds perfectly for me, logging process output exactly as it did before adding the timer. Hopefully this will now be good for merge.

**Pull request merged**